// @feature-hub/history-service/lib/cjs/internal/create-history-service-v3-binder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHistoryServiceV3Binder = void 0;
const browser_consumer_history_1 = require("./browser-consumer-history");
const static_consumer_history_1 = require("./static-consumer-history");
function createHistoryServiceV3(historyKey, consumerHistory, historyMultiplexer) {
    return {
        historyKey,
        history: consumerHistory,
        rootHistory: historyMultiplexer.rootHistory,
        createNewRootLocationForMultipleConsumers: (...consumerLocations) => historyMultiplexer.createNewRootLocationForMultipleConsumers(...consumerLocations),
    };
}
function createBrowserHistoryServiceV3Binding(context, historyMultiplexers, historyKey) {
    const consumerHistory = new browser_consumer_history_1.BrowserConsumerHistory(context, historyKey, historyMultiplexers.browserHistoryMultiplexer);
    return {
        featureService: createHistoryServiceV3(historyKey, consumerHistory, historyMultiplexers.browserHistoryMultiplexer),
        unbind: () => {
            consumerHistory.destroy();
        },
    };
}
function createStaticHistoryServiceV3Binding(context, historyMultiplexers, consumerId) {
    const consumerHistory = new static_consumer_history_1.StaticConsumerHistory(context, consumerId, historyMultiplexers.staticHistoryMultiplexer);
    return {
        featureService: createHistoryServiceV3(consumerId, consumerHistory, historyMultiplexers.staticHistoryMultiplexer),
    };
}
function createHistoryServiceV3Binder(context, historyMultiplexers, options) {
    const { mode, getHistoryKey } = options;
    return (consumerId, consumerName) => {
        const historyKey = getHistoryKey({ consumerId, consumerName });
        return mode === 'browser'
            ? createBrowserHistoryServiceV3Binding(context, historyMultiplexers, historyKey)
            : createStaticHistoryServiceV3Binding(context, historyMultiplexers, historyKey);
    };
}
exports.createHistoryServiceV3Binder = createHistoryServiceV3Binder;
